File: Problems/Two_Pointer/2470.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <algorithm>

using namespace std;

int main()
{
	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	int n;
	cin >> n;

	vector<int> sol(n);

	for (int i = 0; i < n; i++) cin >> sol[i];

	sort(sol.begin(), sol.end());

	auto alkali = sol.begin(), acid = sol.end() - 1;
	pair<int, int> answer(make_pair(0, 0));
	int min = 0x7fffffff;
	while (alkali < acid)
	{
		auto mix = *acid + *alkali;
		auto result = abs(mix);
		if (result <= min)
		{
			min = result;
			answer.first = *alkali;
			answer.second = *acid;
		}
		if (mix == 0) break;
		else if (mix < 0) alkali++;
		else acid--;
	}

	cout << answer.first << " " << answer.second;
	
	return 0;
}

File: Krafton-Jungle/Week-02/2470.py
import sys


n = int(sys.stdin.readline())
sol = list(map(int, sys.stdin.readline().split()))
sol.sort()

left, right = 0, len(sol) - 1

min = 2_100_000_000
alcali = sol[left]
acid = sol[right]
while left < right:
    mix = sol[left] + sol[right]
    result = abs(mix)

    if min >= result:
        min = result
        alcali = sol[left]
        acid = sol[right]
    if mix == 0:
        break
    elif mix < 0:
        left += 1
    else:
        right -= 1

print(alcali, acid)
